src/librustc_mir/interpret/operand.rs

```diff
@@ -289,13 +291,14 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M>
                 let a_ptr = ptr;
                 let b_offset = a_size.abi_align(b.align(self));
                 assert!(b_offset.bytes() > 0); // we later use the offset to test which field to use
-                let b_ptr = ptr.offset(b_offset, self)?.into();
+                let b_ptr = ptr.offset(b_offset, self)?;
                 let a_val = self.memory
                     .get(ptr.alloc_id)?
-                    .read_scalar(self, a_ptr, ptr_align, a_size)?;
+                    .read_scalar(self, a_ptr, a_size)?;
+                self.memory.check_align(b_ptr.into(), b.align(self).min(ptr_align))?;
```

Comment:
This should likely have the same alignment logic as `mplace_field`, which would be `ptr_align.restrict_for_offset(b_offset)`.